### ISSUE TITLE
rpc method listreceivedbyaddress doesn't take address as a string

### ISSUE THREAD

**Issue by RHavar** [bitcoin/bitcoin#14173]
```
bitcoin-cli listreceivedbyaddress 0 true true 2NABbUr9yeRCp1oUCtVmgJF8HGRCo3ifpTT
```

gives an error: `error: Error parsing JSON:2NABbUr9yeRCp1oUCtVmgJF8HGRCo3ifpTT`

as it's trying to parse the address as json. so you'd have to write:
```
bitcoin-cli listreceivedbyaddress 0 true true '"2NABbUr9yeRCp1oUCtVmgJF8HGRCo3ifpTT"'
```
which is clumsy and inconsistent with other RPC methods as well as the help file

**Comment by laanwj** [comment]
can you try the following patch?
```
diff --git a/src/rpc/client.cpp b/src/rpc/client.cpp
index f4d44fc809bffc4c40fbbb5882bcdb2d4a4be720..438d1cb5cb158e0093671093477bec156248d516 100644
--- a/src/rpc/client.cpp
+++ b/src/rpc/client.cpp
@@ -45,7 +45,6 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "listreceivedbyaddress", 0, "minconf" },
     { "listreceivedbyaddress", 1, "include_empty" },
     { "listreceivedbyaddress", 2, "include_watchonly" },
-    { "listreceivedbyaddress", 3, "address_filter" },
     { "listreceivedbylabel", 0, "minconf" },
     { "listreceivedbylabel", 1, "include_empty" },
     { "listreceivedbylabel", 2, "include_watchonly" },
```
(the fourth argument to `listreceivedbyaddress` being called `address_filter` suggests that it can take other JSON expressions, such as a list, but it's really only a JSON string, so there's no reason to treat it differently from other string parameters)

**Comment by RHavar** [comment]
Looks good. Also I think the same thing is happening with the new `getblockstats`

**Comment by MarcoFalke** [comment]
Would be nice to submit the patch with a corresponding test to get this covered?

**Comment by etscrivner** [comment]
Can confirm that something similar is happening with `getblockstats` by making the following changes to the `rpc_getblockstats.py` tests:

```diff
diff --git a/test/functional/rpc_getblockstats.py b/test/functional/rpc_getblockstats.py
index b24bed6ad..0317e63b8 100755
--- a/test/functional/rpc_getblockstats.py
+++ b/test/functional/rpc_getblockstats.py
@@ -51,6 +51,9 @@ class GetblockstatsTest(BitcoinTestFramework):
     def get_stats(self):
         return [self.nodes[0].getblockstats(hash_or_height=self.start_height + i) for i in range(self.max_stat_pos+1)]

+    def get_stats_cli(self):
+        return [self.nodes[0].cli.getblockstats(self.start_height + i) for i in range(self.max_stat_pos+1)]
+
     def generate_test_data(self, filename):
         mocktime = time.time()
         self.nodes[0].generate(101)
@@ -176,5 +179,14 @@ class GetblockstatsTest(BitcoinTestFramework):
         assert_raises_rpc_error(-5, 'Block not found', self.nodes[0].getblockstats,
                                 hash_or_height='000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f')

+        # Test for regressions on #14173
+        cli_stats = self.get_stats_cli()
+        assert_equal(set(cli_stats[0].keys()), set(expected_keys))
+
+        for i in range(self.max_stat_pos+1):
+            blockhash = self.expected_stats[i]['blockhash']
+            stats_by_hash = self.nodes[0].cli.getblockstats(hash_or_height="{}".format(blockhash))
+            assert_equal(stats_by_hash, self.expected_stats[i])
+
 if __name__ == '__main__':
     GetblockstatsTest().main()
```

The fix here is less straightforward since the first parameter can be either a number (in which case it needs to be converted using `ParseNonRFCJSONValue`) or a string (in which case conversion will cause a JSON parsing error)

**Comment by etscrivner** [comment]
I've created a simple fix (#14417) for the initial issue described here. From my initial analysis above it seems as though fixes other similar issues may involve rethinking `ParseNonRFCJsonValue` and its associated code a bit.
